README.md
mlbench: Distributed Machine Learning Benchmark Helm Chart
==========================================================

[![Build Status](https://github.com/mlbench/mlbench-helm/workflows/mlbench-helm/badge.svg?branch=develop)](https://github.com/mlbench/mlbench-helm/workflows/mlbench-helm/badge.svg?branch=develop)
[![DocumentationStatus](https://readthedocs.org/projects/mlbench-docs/badge/?version=latest)](https://mlbench.readthedocs.io/projects/mlbench_helm/en/latest/readme.html?badge=latest)

MLBench is a Benchmarking Framework for Distributed Machine Learning algorithms.

This repository contains the [Helm](https://helm.sh/) Chart for MLBench which is used to install MLBench in a Kubernetes cluster.

For more information refer to the [Helm Documentation](https://mlbench.readthedocs.io/projects/mlbench_helm/en/stable/readme.html)
or the [Main Documentation](https://mlbench.readthedocs.io/)



CHANGELOG.md
# Changelog

## [v3.0.0](https://github.com/mlbench/mlbench-helm/tree/v3.0.0) (2020-12-07)

[Full Changelog](https://github.com/mlbench/mlbench-helm/compare/v2.0.0...v3.0.0)

**Implemented enhancements:**

- Add DIND Setup Script [\#4](https://github.com/mlbench/mlbench-helm/issues/4)
- Add Amazon Cloud setup script [\#3](https://github.com/mlbench/mlbench-helm/issues/3)

**Closed issues:**

- Add integration tests for newer versions of Kubernetes [\#23](https://github.com/mlbench/mlbench-helm/issues/23)
- Add deployment on KIND rather than Minikube [\#21](https://github.com/mlbench/mlbench-helm/issues/21)
- Use of GCloud script [\#19](https://github.com/mlbench/mlbench-helm/issues/19)
- Can not configure NVIDIA on AWS [\#17](https://github.com/mlbench/mlbench-helm/issues/17)
- Migrate to Kubernetes API v1 [\#15](https://github.com/mlbench/mlbench-helm/issues/15)
- Deployment on minikube requires kubernetes 1.15 [\#13](https://github.com/mlbench/mlbench-helm/issues/13)
- Remove obsolete info in `values.yaml` [\#12](https://github.com/mlbench/mlbench-helm/issues/12)
- mlbench worker pods not created [\#11](https://github.com/mlbench/mlbench-helm/issues/11)

**Merged pull requests:**

- Add workflow [\#25](https://github.com/mlbench/mlbench-helm/pull/25) ([ehoelzl](https://github.com/ehoelzl))
- Update to v1 [\#24](https://github.com/mlbench/mlbench-helm/pull/24) ([ehoelzl](https://github.com/ehoelzl))
- Update doc requirements [\#22](https://github.com/mlbench/mlbench-helm/pull/22) ([ehoelzl](https://github.com/ehoelzl))
- Remove AWS and GCloud scripts [\#20](https://github.com/mlbench/mlbench-helm/pull/20) ([ehoelzl](https://github.com/ehoelzl))
- Removes unused entries from values.yaml [\#18](https://github.com/mlbench/mlbench-helm/pull/18) ([Panaetius](https://github.com/Panaetius))
- Switch to eksctl for aws deployment [\#16](https://github.com/mlbench/mlbench-helm/pull/16) ([mmilenkoski](https://github.com/mmilenkoski))
- Add setup script for kind with local registry [\#14](https://github.com/mlbench/mlbench-helm/pull/14) ([mmilenkoski](https://github.com/mmilenkoski))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


run-tests.sh
curl -LO https://storage.googleapis.com/kubernetes-release/release/$KUBECTL_VERSION/bin/linux/amd64/kubectl
chmod +x ./kubectl && sudo mv ./kubectl /usr/local/bin/kubectl
curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3
chmod 700 get_helm.sh
./get_helm.sh
curl https://raw.githubusercontent.com/zlabjp/kubernetes-scripts/master/wait-until-pods-ready > wait_until_pods_ready.sh
chmod +x wait_until_pods_ready.sh
GO111MODULE="on" go get sigs.k8s.io/kind@$KIND_VERSION && kind create cluster --image $KIND_NODE_IMAGE


kubectl cluster-info
kubectl version
helm template travis-test-2 . --set limits.cpu=1000m --set limits.workers=1 --set limits.gpu=0 | kubectl apply --validate -f -
sleep 2
./wait_until_pods_ready.sh 30 2
echo "Deployment successful"

install_on_gke.sh
#!/bin/sh

gcloud container clusters create pytorch-cifar10 --machine-type='n1-standard-2' --accelerator type=nvidia-tesla-k80,count=1 --region europe-west1-b --cluster-version=1.10.9 --num-nodes=17

gcloud container clusters get-credentials pytorch-cifar10 --zone=europe-west1-b

kubectl --namespace kube-system create sa tiller

kubectl create clusterrolebinding tiller --clusterrole cluster-admin --serviceaccount=kube-system:tiller

helm init --wait --service-account tiller

kubectl apply --wait=true -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/stable/nvidia-driver-installer/cos/daemonset-preloaded.yaml

helm upgrade --install --wait --recreate-pods -f myvalues.yaml --timeout 900 --install rel .

export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services rel-mlbench-master)
export NODE_IP=$(gcloud compute instances list|grep $(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}") |awk '{print $5}')
echo http://$NODE_IP:$NODE_PORT
gcloud compute firewall-rules delete --quiet mlbench
gcloud compute firewall-rules create --quiet mlbench --allow tcp:$NODE_PORT,tcp:$NODE_PORT

google_cloud_setup.sh
#!/bin/bash +x
set -e
usage="usage: google_cloud_setup.sh <command>

commands:
    get-credential          Get google credentials
    create-cluster          Create a new cluster
    install-chart           Install the Helm chart
    upgrade-chart           Upgrade (Redeploy) the Helm chart
    uninstall-chart         Delete the Helm release/chart
    delete-cluster          Delete cluster and perform a cleanup
    help                    Show this help

environment variables:
    NUM_NODES               Number of nodes to create in the cluster, default: 2
    PREFIX                  Prefix to add to Cluster and Pod names, default: 'rel'
    MYVALUES_FILE           Path to custom helm chart values file, default: 'myvalues.yaml'

    MACHINE_ZONE            Google Cloud zone, default: 'europe-west1-b'
    MACHINE_TYPE            Google Cloud instance type, default: 'n1-standard-4'
    CLUSTER_VERSION         Kubernetes version, default: 1.11
    DISK_TYPE               Cloud storage type, default: 'pd-standard'
    INSTANCE_DISK_SIZE      Google instance size (GB), default: 50
    NUM_CPUS                Number of CPUs per instance, default: 1
    NUM_GPUS                Number of GPUs per instance, 0 to not use GPU instances, default: 0
    GPU_TYPE                The type of GPU to use, default: 'nvidia-tesla-p100'
    PREEMPTIBLE              Whether to use preemptible instances, default: 0
    "


NUM_NODES=${NUM_NODES:-2}
PREFIX=${PREFIX:-rel}
RELEASE_NAME=${PREFIX}-${NUM_NODES}
CLUSTER_NAME=${PREFIX}-${NUM_NODES}

MACHINE_ZONE=${MACHINE_ZONE:-europe-west1-b}
MYVALUES_FILE=${MYVALUES_FILE:-config.yaml}

MACHINE_TYPE=${MACHINE_TYPE:-n1-standard-4}
CLUSTER_VERSION=${CLUSTER_VERSION:-1.11}
INSTANCE_DISK_SIZE=${INSTANCE_DISK_SIZE:-50}
DISK_TYPE=${DISK_TYPE:-pd-standard}
NUM_CPUS=${NUM_CPUS:-1}
NUM_GPUS=${NUM_GPUS:-0}
GPU_TYPE=${GPU_TYPE:-nvidia-tesla-p100}
PREEMPTIBLE=${PREEMPTIBLE:-0}

MACHINE_ARCHITECTURE=`uname -m`

firewallCheck=`gcloud compute firewall-rules list --filter="name=(${CLUSTER_NAME})"`

if [ ! -f $MYVALUES_FILE ]; then
    echo "Custom Helm values yaml ($MYVALUES_FILE) not found"
    exit 1
fi

function gcloud::check_installed(){
    if ! [ -x "$(command -v gcloud)" ]; then
        echo "Installing Google Cloud SDK"

        if [ ${MACHINE_ARCHITECTURE} == 'x86_64' ]; then
            curl https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-233.0.0-linux-x86_64.tar.gz | tar -xz
        else
            curl https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-233.0.0-linux-x86.tar.gz | tar -xz
        fi

        ./google-cloud-sdk/install.sh
        ./google-cloud-sdk/bin/gcloud init
    fi
}

function helm::check_installed(){
    if ! [ -x "$(command -v helm)" ]; then
        echo "Installing Helm"

        # curl https://raw.githubusercontent.com/helm/helm/master/scripts/get | bash
        source <(curl -s https://raw.githubusercontent.com/helm/helm/master/scripts/get)
    fi
}

function gcloud::get_credential(){
    gcloud::check_installed
    gcloud container clusters get-credentials --zone ${MACHINE_ZONE} ${CLUSTER_NAME}
}

function kube::worker::hostnames(){
    # Get a sequence of names; if you want an array, use additional parentheses ($(kube::worker::hostnames))
    kubectl get pods | grep 'worker' | awk '{print $1}'
}

function kube::worker::ips(){
    # Get a sequence of names; if you want an array, use additional parentheses ($(kube::worker::hostnames))
    kubectl get pods -o wide | grep worker | awk '{print $6}'
}

function chart::upgrade(){
    helm::check_installed

    # Install helm chart
    helm upgrade --wait --recreate-pods -f ${MYVALUES_FILE} \
        --timeout 900 --install ${RELEASE_NAME} . \
        --set limits.workers=$((NUM_NODES-1)) \
        --set limits.gpu=${NUM_GPUS} \
        --set limits.cpu=${NUM_CPUS}
}

function join_by(){
    local IFS="$1";
    shift; echo "$*";
}

function gcloud::cleanup(){
    gcloud::check_installed
    if test -z "$firewallCheck"; then
        echo "firewall rule already deleted"
    else
        gcloud compute firewall-rules delete --quiet ${CLUSTER_NAME}
        echo "firewall rule deleted"
    fi
    gcloud container clusters delete --quiet --zone ${MACHINE_ZONE}  ${CLUSTER_NAME}
}

case $1 in
    create-cluster)
        # Create a CPU cluster
        gcloud::check_installed

        if [ "$PREEMPTIBLE" -gt 0 ]; then
            preemptible=1;
        fi

        if [ "$NUM_GPUS" -gt 0 ]; then
            gcloud container clusters create ${CLUSTER_NAME} \
                --zone=${MACHINE_ZONE} \
                --cluster-version=${CLUSTER_VERSION} \
                --enable-network-policy \
                --machine-type=${MACHINE_TYPE} \
                --accelerator type=${GPU_TYPE},count=${NUM_GPUS} \
                --num-nodes=${NUM_NODES} \
                --disk-type=${DISK_TYPE} \
                --disk-size=${INSTANCE_DISK_SIZE} \
                --scopes=storage-full \
                ${preemptible:+--preemptible}

            kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/stable/nvidia-driver-installer/cos/daemonset-preloaded.yaml
        else
            gcloud container clusters create ${CLUSTER_NAME} \
                --zone=${MACHINE_ZONE} \
                --cluster-version=${CLUSTER_VERSION} \
                --enable-network-policy \
                --machine-type=${MACHINE_TYPE} \
                --num-nodes=${NUM_NODES} \
                --disk-type=${DISK_TYPE} \
                --disk-size=${INSTANCE_DISK_SIZE} \
                --scopes=storage-full\
                ${preemptible:+--preemptible}
        fi


        # Get credential of the cluster
        gcloud container clusters get-credentials --zone ${MACHINE_ZONE} ${CLUSTER_NAME}

        kubectl --namespace kube-system create sa tiller

        kubectl create clusterrolebinding tiller --clusterrole cluster-admin --serviceaccount=kube-system:tiller

        # Initialize helm to install charts
        helm::check_installed
        helm init --wait --service-account tiller
        ;;

    cleanup-cluster )
        gcloud::check_installed
        if test -z "$firewallCheck"; then
            echo "firewall rule already deleted"
        else
            gcloud compute firewall-rules delete --quiet ${CLUSTER_NAME}
            echo "firewall rule deleted"
        fi
        gcloud container clusters delete --quiet --zone ${MACHINE_ZONE}  ${CLUSTER_NAME}
        ;;

    install-chart)
        chart::upgrade

        # setup firewall
        gcloud::check_installed
        export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services ${RELEASE_NAME}-mlbench-master)

        if [ "$PREEMPTIBLE" -gt 0 ]; then
            export NODE_IP=$(gcloud compute instances list|grep $(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}") |awk '{print $6}')
        else
            export NODE_IP=$(gcloud compute instances list|grep $(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}") |awk '{print $5}')
        fi

        gcloud compute firewall-rules create --quiet ${CLUSTER_NAME} --allow tcp:$NODE_PORT,tcp:$NODE_PORT
        echo "You can access MLBench at the following URL:"
        echo http://$NODE_IP:$NODE_PORT
        ;;

    upgrade-chart)
        chart::upgrade
        ;;


    uninstall-chart)
        gcloud::check_installed
        helm::check_installed
        helm delete --purge ${RELEASE_NAME}
        gcloud compute firewall-rules delete --quiet ${CLUSTER_NAME}
        ;;

    delete-cluster)
        gcloud::cleanup
        ;;

    get-credential)
        gcloud::get_credential
        ;;
    help)
        echo "$usage"
        ;;
    *)
        printf "illegal option: %s\n" "$1" >&2
        echo "$usage" >&2

esac


kind-with-registry.sh
#!/bin/sh
set -o errexit

# desired cluster name; default is "kind"
KIND_CLUSTER_NAME="${KIND_CLUSTER_NAME:-kind}"

# create registry container unless it already exists
reg_name='kind-registry'
reg_port='5000'
running="$(docker inspect -f '{{.State.Running}}' "${reg_name}" 2>/dev/null || true)"
if [ "${running}" != 'true' ]; then
  docker run \
    -d --restart=always -p "${reg_port}:5000" --name "${reg_name}" \
    registry:2
fi
reg_ip="$(docker inspect -f '{{.NetworkSettings.IPAddress}}' "${reg_name}")"

# create a cluster with the local registry enabled in containerd
cat <<EOF | kind create cluster --name "${KIND_CLUSTER_NAME}" --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
containerdConfigPatches: 
- |-
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${reg_port}"]
    endpoint = ["http://${reg_ip}:${reg_port}"]
kubeadmConfigPatches:
- |
  kind: ClusterConfiguration
  metadata:
    name: config
  apiServer:
    extraArgs:
      "runtime-config": "apps/v1beta1=true,apps/v1beta2=true,extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true"
nodes:
- role: control-plane
- role: worker
- role: worker
EOF


aws_setup.sh
#!/bin/bash +x
set -e
usage="usage: aws_setup.sh <command>

commands:
    get-credential          Get aws credentials
    create-cluster          Create a new cluster
    install-chart           Install the Helm chart
    upgrade-chart           Upgrade (Redeploy) the Helm chart
    uninstall-chart         Delete the Helm release/chart
    delete-cluster          Delete cluster and perform a cleanup
    help                    Show this help

environment variables:
    NUM_NODES               Number of nodes to create in the cluster, default: 2
    PREFIX                  Prefix to add to Cluster and Pod names, default: 'rel'
    MYVALUES_FILE           Path to custom helm chart values file, default: 'myvalues.yaml'

    MACHINE_ZONE            Google Cloud zone, default: 'europe-west1-b'
    MACHINE_TYPE            Google Cloud instance type, default: 't2.medium'
    CLUSTER_VERSION         Kubernetes version, default: 1.10
    DISK_TYPE               Cloud storage type, default: 'pd-standard'
    INSTANCE_DISK_SIZE      Google instance size (GB), default: 50
    NUM_CPUS                Number of CPUs per instance, default: 1
    NUM_GPUS                Number of GPUs per instance, 0 to not use GPU instances, default: 0
    GPU_TYPE                The type of GPU to use, default: 'nvidia-tesla-p100'

    "

NUM_NODES=${NUM_NODES:-3}
PREFIX=${PREFIX:-rel}
RELEASE_NAME=${PREFIX}-${NUM_NODES}
CLUSTER_NAME=${PREFIX}-${NUM_NODES}b
DEV_NAME=${DEV_NAME:-/dev/sdh}

MACHINE_ZONE=${MACHINE_ZONE:-us-east-1}
MYVALUES_FILE=${MYVALUES_FILE:-values.yaml}
KEY_NAME=${KEY_NAME:-MyKey}
SECURITY_GROUP=${SECURITY_GROUP:-EC2SecurityGroup}

IMAGE_ID=${IMAGE_ID:-ami-075b44448d2276521}
MACHINE_TYPE=${MACHINE_TYPE:-t2.medium}
KUBERNETES_VERSION=${KUBERNETES_VERSION:-1.15}
CLUSTER_VERSION=${CLUSTER_VERSION:-1.11.7-gke.6}
INSTANCE_DISK_SIZE=${INSTANCE_DISK_SIZE:-50}
DISK_TYPE=${DISK_TYPE:-pd-standard}
NUM_CPUS=${NUM_CPUS:-1}
NUM_GPUS=${NUM_GPUS:-0}
GPU_TYPE=${GPU_TYPE:-nvidia-tesla-p100}



MACHINE_ARCHITECTURE=`uname -m`

if [ ! -f $MYVALUES_FILE ]; then
    echo "Custom Helm values yaml ($MYVALUES_FILE) not found"
    exit 1
fi

function kubectl::check_installed(){
    if ! [ -x "$(command -v kubectl)" ]; then
        echo "Installing kubectl"
        curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl
        chmod +x ./kubectl
        #should use sudo? what if no admin priviledges? easier to install with 'sudo snap install kubectl --classic'
        sudo mv ./kubectl /usr/local/bin/kubectl
    fi
}

function aws::check_installed(){
    if ! [ -x "$(command -v aws)" ]; then
        echo "Installing AWS CLI"

    	# download and install aws cli
	curl "https://s3.amazonaws.com/aws-cli/awscli-bundle.zip" -o "awscli-bundle.zip"
	unzip awscli-bundle.zip
	./awscli-bundle/install -b ~/.local/bin/aws
	echo "export PATH=\$PATH:\$HOME/.local/bin" >> $HOME/.bashrc
    export PATH=$PATH:$HOME/.local/bin
	source $HOME/.bashrc

    # configure access credentials, may require use input
    aws configure

    # generate access credentials
    aws iam create-group --group-name kops
    aws iam attach-group-policy --policy-arn arn:aws:iam::aws:policy/AmazonEC2FullAccess --group-name kops

    aws iam attach-group-policy --policy-arn arn:aws:iam::aws:policy/AmazonRoute53FullAccess --group-name kops
    aws iam attach-group-policy --policy-arn arn:aws:iam::aws:policy/AmazonS3FullAccess --group-name kops
    aws iam attach-group-policy --policy-arn arn:aws:iam::aws:policy/IAMFullAccess --group-name kops
    aws iam attach-group-policy --policy-arn arn:aws:iam::aws:policy/AmazonVPCFullAccess --group-name kops

    aws iam create-user --user-name kops
    aws iam add-user-to-group --user-name kops --group-name kops
    aws iam create-access-key --user-name kops

    # export env variables for kops to use
    export AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id)
    export AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key)

    # cleanup installation package
    rm -r ./awscli-bundle/
    rm ./awscli-bundle.zip
    fi
}

function eksctl::check_installed(){
    if ! [ -x "$(command -v eksctl)" ]; then
        echo "Installing eksctl"

    	# download and install eksctl
	curl --silent --location "https://github.com/weaveworks/eksctl/releases/latest/download/eksctl_$(uname -s)_amd64.tar.gz" | tar xz -C /tmp
	sudo mv /tmp/eksctl /usr/local/bin
    fi
}


function helm::check_installed(){
    if ! [ -x "$(command -v helm)" ]; then
        echo "Installing Helm"

        # curl https://raw.githubusercontent.com/helm/helm/master/scripts/get | bash
        source <(curl -s https://raw.githubusercontent.com/helm/helm/master/scripts/get)
    fi
}

function aws::get_credential(){
    aws::check_installed
    kubectl::check_installed
}

function kube::worker::hostnames(){
    # Get a sequence of names; if you want an array, use additional parentheses ($(kube::worker::hostnames))
    kubectl get pods | grep 'worker' | awk '{print $1}'
}

function kube::worker::ips(){
    # Get a sequence of names; if you want an array, use additional parentheses ($(kube::worker::hostnames))
    kubectl get pods -o wide | grep worker | awk '{print $6}'
}

function chart::upgrade(){
    helm::check_installed

    # Install helm chart
    helm upgrade --wait --recreate-pods -f ${MYVALUES_FILE} \
        --timeout 900s --install ${RELEASE_NAME} . \
        --set limits.workers=$((NUM_NODES-1)) \
        --set limits.gpu=${NUM_GPUS} \
        --set limits.cpu=${NUM_CPUS}
}

function join_by(){
    local IFS="$1";
    shift; echo "$*";
}

function aws::cleanup(){
    aws::check_installed
    eksctl::check_installed
    eksctl delete cluster --name "${CLUSTER_NAME}"
}

case $1 in
    create-cluster)
        # Create a CPU cluster
        aws::check_installed
	eksctl::check_installed

        if [ ! -d ~/.ssh ]; then
            ssh-keygen
        fi

        eksctl create cluster \
		--name "${CLUSTER_NAME}" \
		--version ${KUBERNETES_VERSION} \
		--region ${MACHINE_ZONE} \
		--nodegroup-name standard-workers \
		--node-type "${MACHINE_TYPE}" \
		--nodes ${NUM_NODES} \
		--nodes-min $((NUM_NODES-1)) \
		--nodes-max $((NUM_NODES+1)) \
		--ssh-access \
		--managed

        if [ "$NUM_GPUS" -gt 0 ]; then
		echo "Installing NVIDIA drivers"
		kubectl create -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/1.0.0-beta6/nvidia-device-plugin.yml
        fi
        ;;

    cleanup-cluster )
        aws::check_installed
	eksctl::check_installed
        eksctl delete cluster --name "${CLUSTER_NAME}"
        ;;

    install-chart)
        chart::upgrade

        # setup firewall
        aws::check_installed
        export NODE_PORT=$(kubectl get --namespace default -o jsonpath="{.spec.ports[0].nodePort}" services ${RELEASE_NAME}-mlbench-master)
        export NODE_IP=$(kubectl get nodes --namespace default -o jsonpath="{.items[0].status.addresses[0].address}")
        export MAIN_MACHINE_ZONE=$(echo $MACHINE_ZONE | sed -e 's/\([a-z]\)*$//g')
        export GROUP_ID=$(aws ec2 describe-instances --region ${MAIN_MACHINE_ZONE} --filter Name=private-ip-address,Values=${NODE_IP} --query 'Reservations[].Instances[].[SecurityGroups][0][0][0].GroupId')
        export GROUP_ID=${GROUP_ID:1:-1}
        export NODE_IP=$(aws ec2 describe-instances --region ${MAIN_MACHINE_ZONE} --filter Name=private-ip-address,Values=${NODE_IP} --query 'Reservations[].Instances[].[InstanceId,PublicIpAddress][0][1]')
        export NODE_IP=${NODE_IP:1:-1}
        #export VPC_ID=$(aws ec2 describe-instances --region ${MAIN_MACHINE_ZONE} --filter Name=private-ip-address,Values=${NODE_IP} --query 'Reservations[].Instances[].[VpcId][0][0]')
        #export VPC_ID=${VPC_ID:1:-1}
        aws ec2 authorize-security-group-ingress --region ${MAIN_MACHINE_ZONE} --group-id ${GROUP_ID} --protocol tcp --port ${NODE_PORT} --cidr 0.0.0.0/0

        echo "You can access MLBench at the following URL:"
        echo http://$NODE_IP:$NODE_PORT
        ;;

    upgrade-chart)
        chart::upgrade
        ;;


    uninstall-chart)
        aws::check_installed
        helm::check_installed
        helm uninstall ${RELEASE_NAME}
        ;;

    delete-cluster)
        aws::cleanup
        ;;

    get-credential)
        aws::get_credential
        ;;
    help)
        echo "$usage"
        ;;
    *)
        printf "illegal option: %s\n" "$1" >&2
        echo "$usage" >&2

esac
